src/main.rs

```diff
@@ -467,18 +483,110 @@ async fn forward_request(
     Ok(response)
 }
 
-fn validate_body(
-    certificate: &[u8],
-    tree: &[u8],
+fn validate(
+    headers_data: &HeadersData,
     canister_id: &Principal,
     agent: &Agent,
     uri: &Uri,
     response_body: &[u8],
     logger: slog::Logger,
+) -> Result<(), String> {
+    let body_sha = if let Some(body_sha) = decode_body(response_body, headers_data.encoding.clone())
+    {
+        body_sha
+    } else {
+        return Err("Body could not be decoded".into());
+    };
+
+    let body_valid = match (
+        headers_data.certificate.as_ref(),
+        headers_data.tree.as_ref(),
+    ) {
+        (Some(Ok(certificate)), Some(Ok(tree))) => match validate_body(
+            Certificates { certificate, tree },
+            canister_id,
+            agent,
+            uri,
+            &body_sha,
+            logger.clone(),
+        ) {
+            Ok(valid) => {
+                if valid {
+                    Ok(())
+                } else {
+                    Err("Body does not pass verification".to_string())
+                }
+            }
+            Err(e) => Err(format!("Certificate validation failed: {}", e)),
+        },
+        (Some(_), _) | (_, Some(_)) => Err("Body does not pass verification".to_string()),
+        // Canisters don't have to provide certified variables
```

Comment:
Add TODO for explicit opt out and attach Jira ticket (let's discuss this).  Current thought is that this would be via metadata for grandfathered in canisters.

Reply:
@Daniel-Bloom-dfinity can you help with TODO and JIRA ticket?

Reply:
Done